fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingMutuallyByRecords.java

```diff
@@ -0,0 +1,529 @@
+    }
+
+    @Override
+    @Nonnull
+    IndexBuildProto.IndexBuildIndexingStamp getIndexingTypeStamp(FDBRecordStore store) {
+        if (myIndexingTypeStamp == null) {
+            myIndexingTypeStamp = compileIndexingTypeStamp(common.getTargetIndexesNames());
+        }
+        return myIndexingTypeStamp;
+    }
+
+    @Nonnull
+    private static IndexBuildProto.IndexBuildIndexingStamp compileIndexingTypeStamp(List<String> targetIndexes) {
+
+        if (targetIndexes.isEmpty()) {
+            throw new ValidationException("No target index was set");
+        }
+        return IndexBuildProto.IndexBuildIndexingStamp.newBuilder()
+                .setMethod(IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS)
+                .addAllTargetIndex(targetIndexes)
+                .build();
+    }
+
+    private static boolean areTheyAllIdempotent(@Nonnull FDBRecordStore store, List<Index> targetIndexes) {
+        return targetIndexes.stream()
+                .allMatch(targetIndex -> store.getIndexMaintainer(targetIndex).isIdempotent());
+    }
+
+    @Override
+    List<Object> indexingLogMessageKeyValues() {
+        return Arrays.asList(
+                LogMessageKeys.INDEXING_METHOD, "mutual multi target by records",
+                LogMessageKeys.TARGET_INDEX_NAME, common.getTargetIndexesNames()
+        );
+    }
+
+    private List<Tuple> getPrimaryKeyBoundaries(@Nonnull FDBRecordStore store) {
+        TupleRange tupleRange = common.computeRecordsRange();
+        store.getContext().getReadVersion(); // for instrumentation reasons
+        List<Tuple> boundaries;
+        try (RecordCursor<Tuple> cursor = store.getPrimaryKeyBoundaries(tupleRange)) {
+            boundaries = cursor.asList().join();
+        }
+
+        if (boundaries == null) {
+            boundaries = new ArrayList<>();
+        }
+
+        // Add the two endpoints to the range
+        if (tupleRange == null) {
+            // Here: make sure that the boundaries cover everything
+            boundaries.add(0, null);
+            boundaries.add(null);
+        } else {
+            // Here: add the endpoints, unless they are already included
+            if (boundaries.isEmpty() || tupleRange.getLow() == null || tupleRange.getLow().compareTo(boundaries.get(0)) < 0) {
+                boundaries.add(0, tupleRange.getLow());
+            }
+            if (tupleRange.getHigh() == null || tupleRange.getHigh().compareTo(boundaries.get(boundaries.size() - 1)) > 0) {
+                boundaries.add(tupleRange.getHigh());
+            }
+        }
+
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(KeyValueLogMessage.of("got boundaries",
+                    LogMessageKeys.INDEX_NAME, common.getTargetIndexesNames(),
+                    LogMessageKeys.RANGE, tupleRange,
+                    LogMessageKeys.KEY_COUNT, boundaries.size()));
+        }
+
+        return boundaries;
+    }
+
+    private int getPrimeStep(int size, Random rn) {
```

Comment:
There's some tricky logic in here around taking a list, and iterating over the list with a step. Perhaps it would make sense to pull that out into a helper class, so that it could be unit-tested.

Reply:
After thinking about it, it seems that any reasonable parameters test is already covered - this way or another - by the existing indexing tests ("weird" boundaries or not). Especially when considering the random nature of  fragmentFirst & fragmentStep.

Reply:
Yes, but the randomness would mean that the test could be flaky, rather than catching a bad situation reliably. By having a test that makes sure that using `getPrimeStep` and then `fragmentPlusPlus` actually visits all values in the fragments. As a unit test, you could pretty efficiently do 100 repetitions for size=1 to size=30